logger/logger.go

```diff
@@ -86,6 +87,20 @@ func LogrusLogLevelFrom(configLogLevel string) logrus.Level {
 	return logLevel
 }
 
+func backtrace() string {
```

Comment:
I would like to ask your opinion on returning the array of strings directly instead of joining them, and also about including the caller function in a very "Java"ey style:

```diff
-  func backtrace() string
+ func backtrace() []string {
	var filenames []string

	pc := make([]uintptr, 1000)
	n := runtime.Callers(0, pc)
	frames := runtime.CallersFrames(pc[:n])

	for frame, isNextFrameValid := frames.Next(); isNextFrameValid; frame, isNextFrameValid = frames.Next() {
-               filenames = append(filenames, fmt.Sprintf("%s:%d", frame.File, frame.Line))
+		filenames = append(filenames, fmt.Sprintf("%s(%s:%d)", frame.Function, frame.File, frame.Line))
	}

-       return strings.Join(filenames, ";")
+	return filenames
}
```

That way we would get the following —inlined— backtrace object:


```json
    "backtrace": [
        "runtime.Callers(/usr/lib/go-1.17/src/runtime/extern.go:227)",
        "github.com/RedHatInsights/sources-api-go/logger.backtrace(/home/mikelalejobr/development/sources/sources-api-go/logger/logger.go:94)",
        "github.com/RedHatInsights/sources-api-go/logger.(*CustomLoggerFormatter).Format(/home/mikelalejobr/development/sources/sources-api-go/logger/logger.go:156)",
        "github.com/sirupsen/logrus.(*Entry).write(/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go:279)",
        "github.com/sirupsen/logrus.(*Entry).log(/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go:251)",
        "github.com/sirupsen/logrus.(*Entry).Log(/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go:293)",
        "github.com/sirupsen/logrus.(*Logger).Log(/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/logger.go:198)",
        "github.com/sirupsen/logrus.(*Logger).Error(/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/logger.go:238)",
        "github.com/RedHatInsights/sources-api-go/logger.(*CustomGORMLogger).Error(/home/mikelalejobr/development/sources/sources-api-go/logger/gorm_logger.go:33)",
        "gorm.io/gorm.Open(/home/mikelalejobr/go/pkg/mod/gorm.io/gorm@v1.23.2/gorm.go:206)",
        "github.com/RedHatInsights/sources-api-go/dao.Init(/home/mikelalejobr/development/sources/sources-api-go/dao/db.go:85)",
        "main.main(/home/mikelalejobr/development/sources/sources-api-go/main.go:30)",
        "runtime.main(/usr/lib/go-1.17/src/runtime/proc.go:255)"
    ],
```

---

I also played with the possibility of using a more structured output:

```go
type report struct {
	Filename string `json:"filename"`
	Function string `json:"function"`
	Line     int    `json:"line"`
}
```

But to be honest I don't really think we will ever need this level of filtering detail:

```json
    "backtrace": [
        {
            "function": "runtime.Callers",
            "filename": "/usr/lib/go-1.17/src/runtime/extern.go",
            "line": 227
        },
        {
            "function": "github.com/RedHatInsights/sources-api-go/logger.backtrace",
            "filename": "/home/mikelalejobr/development/sources/sources-api-go/logger/logger.go",
            "line": 100
        },
        {
            "function": "github.com/RedHatInsights/sources-api-go/logger.(*CustomLoggerFormatter).Format",
            "filename": "/home/mikelalejobr/development/sources/sources-api-go/logger/logger.go",
            "line": 171
        },
        {
            "function": "github.com/sirupsen/logrus.(*Entry).write",
            "filename": "/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go",
            "line": 279
        },
        {
            "function": "github.com/sirupsen/logrus.(*Entry).log",
            "filename": "/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go",
            "line": 251
        },
        {
            "function": "github.com/sirupsen/logrus.(*Entry).Log",
            "filename": "/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/entry.go",
            "line": 293
        },
        {
            "function": "github.com/sirupsen/logrus.(*Logger).Log",
            "filename": "/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/logger.go",
            "line": 198
        },
        {
            "function": "github.com/sirupsen/logrus.(*Logger).Error",
            "filename": "/home/mikelalejobr/go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/logger.go",
            "line": 238
        },
        {
            "function": "github.com/RedHatInsights/sources-api-go/logger.(*CustomGORMLogger).Error",
            "filename": "/home/mikelalejobr/development/sources/sources-api-go/logger/gorm_logger.go",
            "line": 33
        },
        {
            "function": "gorm.io/gorm.Open",
            "filename": "/home/mikelalejobr/go/pkg/mod/gorm.io/gorm@v1.23.2/gorm.go",
            "line": 206
        },
        {
            "function": "github.com/RedHatInsights/sources-api-go/dao.Init",
            "filename": "/home/mikelalejobr/development/sources/sources-api-go/dao/db.go",
            "line": 85
        },
        {
            "function": "main.main",
            "filename": "/home/mikelalejobr/development/sources/sources-api-go/main.go",
            "line": 30
        },
        {
            "function": "runtime.main",
            "filename": "/usr/lib/go-1.17/src/runtime/proc.go",
            "line": 255
        }
    ],
```

This would allow much easier filtering in Kibana, but to be honest are we ever going to filter log messages by their error function, filename, or line number? I don't think so...

Reply:
I like inlined backtrace object and I added it 👍 

Yes obviously such filtering is not needed. Nested version of logs looks very nice but I think for our better navigation in backtrace is useful firstly mentioned version - so If you are OK with it I would keep inline version.

Reply:
Absolutely! I also think that the first option is more than enough. Thanks!